File: backend/app/routers/egresos.py
"""
Endpoints de egresos: tipos de gasto, gastos y pagos.

Ninguno es un comprobante electrónico —el SRI no los recibe— pero sin ellos no
se sabe cuánto se gastó, y el formulario 104 declara también las compras.

La distinción entre **gasto** y **egreso** es deliberada y vale la pena
recordarla: el gasto es la obligación (llegó la factura del arriendo) y el
egreso es la salida de dinero (se pagó el arriendo). No coinciden en el tiempo
ni en el importe, así que se registran por separado.
"""

from __future__ import annotations

from datetime import date

from fastapi import APIRouter, Depends, HTTPException, Query, Response
from sqlalchemy import func, or_, select
from sqlalchemy.orm import Session

from ..base_datos import obtener_sesion
from ..esquemas import (
    EgresoEntrada,
    EgresoSalida,
    GastoEntrada,
    GastoSalida,
    ResumenEgresos,
    TipoGastoEntrada,
    TipoGastoSalida,
)
from ..modelos_db import Egreso, Gasto, Receptor, TipoGasto
from ..sri.modelos import redondear

router = APIRouter(prefix="/egresos", tags=["egresos"])


# --------------------------------------------------------------------------
# Tipos de gasto
# --------------------------------------------------------------------------


@router.get("/tipos", response_model=list[TipoGastoSalida])
def listar_tipos(
    sesion: Session = Depends(obtener_sesion),
    incluir_inactivos: bool = False,
):
    consulta = select(TipoGasto).order_by(TipoGasto.nombre)
    if not incluir_inactivos:
        consulta = consulta.where(TipoGasto.estado == "Activo")
    return sesion.scalars(consulta).all()


@router.post("/tipos", response_model=TipoGastoSalida, status_code=201)
def crear_tipo(datos: TipoGastoEntrada, sesion: Session = Depends(obtener_sesion)):
    repetido = sesion.scalar(select(TipoGasto).where(TipoGasto.nombre == datos.nombre))
    if repetido:
        raise HTTPException(409, f"Ya existe un tipo de gasto llamado «{datos.nombre}».")

    tipo = TipoGasto(**datos.model_dump())
    sesion.add(tipo)
    sesion.commit()
    sesion.refresh(tipo)
    return tipo


@router.put("/tipos/{tipo_id}", response_model=TipoGastoSalida)
def actualizar_tipo(
    tipo_id: int, datos: TipoGastoEntrada, sesion: Session = Depends(obtener_sesion)
):
    tipo = sesion.get(TipoGasto, tipo_id)
    if tipo is None:
        raise HTTPException(404, "Tipo de gasto no encontrado.")

    repetido = sesion.scalar(
        select(TipoGasto).where(TipoGasto.nombre == datos.nombre, TipoGasto.id != tipo_id)
    )
    if repetido:
        raise HTTPException(409, f"Ya existe un tipo de gasto llamado «{datos.nombre}».")

    for campo, valor in datos.model_dump().items():
        setattr(tipo, campo, valor)

    sesion.commit()
    sesion.refresh(tipo)
    return tipo


@router.delete("/tipos/{tipo_id}", status_code=204)
def desactivar_tipo(tipo_id: int, sesion: Session = Depends(obtener_sesion)):
    """
    Se desactiva, no se borra.

    Los gastos ya registrados apuntan al tipo; borrarlo los dejaría sin
    categoría y descuadraría los reportes de meses ya cerrados.
    """
    tipo = sesion.get(TipoGasto, tipo_id)
    if tipo is None:
        raise HTTPException(404, "Tipo de gasto no encontrado.")

    tipo.estado = "Inactivo"
    sesion.commit()


# --------------------------------------------------------------------------
# Gastos
# --------------------------------------------------------------------------


@router.get("/gastos", response_model=list[GastoSalida])
def listar_gastos(
    respuesta: Response,
    sesion: Session = Depends(obtener_sesion),
    buscar: str | None = None,
    tipo_id: int | None = None,
    estado_pago: str | None = None,
    desde: date | None = None,
    hasta: date | None = None,
    pagina: int = Query(1, ge=1),
    tamano: int = Query(20, ge=1, le=200),
):
    consulta = select(Gasto)

    if buscar:
        patron = f"%{buscar.lower()}%"
        consulta = consulta.where(
            or_(
                func.lower(Gasto.concepto).like(patron),
                func.lower(Gasto.proveedor_razon_social).like(patron),
                func.lower(Gasto.documento).like(patron),
            )
        )
    if tipo_id:
        consulta = consulta.where(Gasto.tipo_id == tipo_id)
    if estado_pago:
        consulta = consulta.where(Gasto.estado_pago == estado_pago)
    if desde:
        consulta = consulta.where(Gasto.fecha >= desde)
    if hasta:
        consulta = consulta.where(Gasto.fecha <= hasta)

    total = sesion.scalar(select(func.count()).select_from(consulta.subquery())) or 0
    respuesta.headers["X-Total-Registros"] = str(total)

    consulta = (
        consulta.order_by(Gasto.fecha.desc(), Gasto.id.desc())
        .offset((pagina - 1) * tamano)
        .limit(tamano)
    )
    return sesion.scalars(consulta).all()


def _validar_tipo(sesion: Session, tipo_id: int | None) -> TipoGasto | None:
    if tipo_id is None:
        return None
    tipo = sesion.get(TipoGasto, tipo_id)
    if tipo is None:
        raise HTTPException(404, "El tipo de gasto indicado no existe.")
    return tipo


def _validar_proveedor(sesion: Session, proveedor_id: int | None) -> Receptor | None:
    if proveedor_id is None:
        return None
    proveedor = sesion.get(Receptor, proveedor_id)
    if proveedor is None:
        raise HTTPException(404, "El proveedor indicado no existe.")
    # Nada impedía registrar un gasto contra un cliente: el desplegable del
    # formulario lista todos los receptores, no solo los que tienen ese rol.
    if proveedor.rol != "Proveedor":
        raise HTTPException(
            422,
            f"{proveedor.razon_social} está registrado como {proveedor.rol}, no "
            "como Proveedor. Cambia su rol en Receptores o elige otro.",
        )
    return proveedor


@router.post("/gastos", response_model=GastoSalida, status_code=201)
def crear_gasto(datos: GastoEntrada, sesion: Session = Depends(obtener_sesion)):
    _validar_tipo(sesion, datos.tipo_id)
    proveedor = _validar_proveedor(sesion, datos.proveedor_id)

    gasto = Gasto(
        fecha=datos.fecha or date.today(),
        concepto=datos.concepto,
        tipo_id=datos.tipo_id,
        proveedor_id=datos.proveedor_id,
        # Se copia el nombre además del id: si mañana se desactiva el
        # proveedor, el gasto histórico sigue diciendo a quién se le pagó.
        proveedor_razon_social=proveedor.razon_social if proveedor else "",
        proveedor_identificacion=proveedor.identificacion if proveedor else "",
        documento=datos.documento,
        fecha_documento=datos.fecha_documento,
        autorizacion_proveedor=datos.autorizacion_proveedor,
        subtotal=datos.subtotal,
        iva=datos.iva,
        codigo_iva=datos.codigo_iva,
        total=redondear(datos.subtotal + datos.iva),
        estado_pago=datos.estado_pago,
        observacion=datos.observacion,
    )

    sesion.add(gasto)
    sesion.commit()
    sesion.refresh(gasto)
    return gasto


@router.get("/gastos/{gasto_id}", response_model=GastoSalida)
def obtener_gasto(gasto_id: int, sesion: Session = Depends(obtener_sesion)):
    gasto = sesion.get(Gasto, gasto_id)
    if gasto is None:
        raise HTTPException(404, "Gasto no encontrado.")
    return gasto


@router.put("/gastos/{gasto_id}", response_model=GastoSalida)
def actualizar_gasto(
    gasto_id: int, datos: GastoEntrada, sesion: Session = Depends(obtener_sesion)
):
    gasto = sesion.get(Gasto, gasto_id)
    if gasto is None:
        raise HTTPException(404, "Gasto no encontrado.")

    _validar_tipo(sesion, datos.tipo_id)
    proveedor = _validar_proveedor(sesion, datos.proveedor_id)

    gasto.fecha = datos.fecha or gasto.fecha
    gasto.concepto = datos.concepto
    gasto.tipo_id = datos.tipo_id
    gasto.proveedor_id = datos.proveedor_id
    gasto.proveedor_razon_social = proveedor.razon_social if proveedor else ""
    gasto.proveedor_identificacion = proveedor.identificacion if proveedor else ""
    gasto.documento = datos.documento
    gasto.fecha_documento = datos.fecha_documento
    gasto.autorizacion_proveedor = datos.autorizacion_proveedor
    gasto.subtotal = datos.subtotal
    gasto.iva = datos.iva
    gasto.codigo_iva = datos.codigo_iva
    gasto.total = redondear(datos.subtotal + datos.iva)
    gasto.estado_pago = datos.estado_pago
    gasto.observacion = datos.observacion

    sesion.commit()
    sesion.refresh(gasto)
    return gasto


@router.delete("/gastos/{gasto_id}", status_code=204)
def eliminar_gasto(gasto_id: int, sesion: Session = Depends(obtener_sesion)):
    gasto = sesion.get(Gasto, gasto_id)
    if gasto is None:
        raise HTTPException(404, "Gasto no encontrado.")

    pagado = sesion.scalar(select(func.count(Egreso.id)).where(Egreso.gasto_id == gasto_id))
    if pagado:
        raise HTTPException(
            409,
            "Este gasto tiene pagos registrados. Anula primero los egresos "
            "asociados para no dejar la caja descuadrada.",
        )

    sesion.delete(gasto)
    sesion.commit()


# --------------------------------------------------------------------------
# Egresos (pagos)
# --------------------------------------------------------------------------


@router.get("", response_model=list[EgresoSalida])
def listar_egresos(
    respuesta: Response,
    sesion: Session = Depends(obtener_sesion),
    buscar: str | None = None,
    forma_pago: str | None = None,
    estado: str | None = None,
    desde: date | None = None,
    hasta: date | None = None,
    pagina: int = Query(1, ge=1),
    tamano: int = Query(20, ge=1, le=200),
):
    consulta = select(Egreso)

    if buscar:
        patron = f"%{buscar.lower()}%"
        consulta = consulta.where(
            or_(
                func.lower(Egreso.concepto).like(patron),
                func.lower(Egreso.beneficiario).like(patron),
                func.lower(func.coalesce(Egreso.referencia, "")).like(patron),
            )
        )
    if forma_pago:
        consulta = consulta.where(Egreso.forma_pago == forma_pago)
    if estado:
        consulta = consulta.where(Egreso.estado == estado)
    if desde:
        consulta = consulta.where(Egreso.fecha >= desde)
    if hasta:
        consulta = consulta.where(Egreso.fecha <= hasta)

    total = sesion.scalar(select(func.count()).select_from(consulta.subquery())) or 0
    respuesta.headers["X-Total-Registros"] = str(total)

    consulta = (
        consulta.order_by(Egreso.fecha.desc(), Egreso.id.desc())
        .offset((pagina - 1) * tamano)
        .limit(tamano)
    )
    return sesion.scalars(consulta).all()


@router.post("", response_model=EgresoSalida, status_code=201)
def crear_egreso(datos: EgresoEntrada, sesion: Session = Depends(obtener_sesion)):
    gasto = None
    if datos.gasto_id is not None:
        gasto = sesion.get(Gasto, datos.gasto_id)
        if gasto is None:
            raise HTTPException(404, "El gasto indicado no existe.")

    egreso = Egreso(
        fecha=datos.fecha or date.today(),
        concepto=datos.concepto,
        beneficiario=datos.beneficiario,
        monto=datos.monto,
        forma_pago=datos.forma_pago,
        cuenta_id=datos.cuenta_id,
        referencia=datos.referencia,
        gasto_id=datos.gasto_id,
        observacion=datos.observacion,
    )
    sesion.add(egreso)

    # Si el pago cubre el gasto entero, este deja de estar pendiente. Se
    # compara contra la suma de todos sus pagos, no solo contra este.
    if gasto is not None:
        pagado = sesion.scalar(
            select(func.coalesce(func.sum(Egreso.monto), 0)).where(
                Egreso.gasto_id == gasto.id, Egreso.estado != "Anulado"
            )
        )
        if redondear(pagado or 0) + datos.monto >= gasto.total:
            gasto.estado_pago = "Pagado"
        else:
            gasto.estado_pago = "Parcial"

    sesion.commit()
    sesion.refresh(egreso)
    return egreso


@router.get("/{egreso_id}", response_model=EgresoSalida)
def obtener_egreso(egreso_id: int, sesion: Session = Depends(obtener_sesion)):
    egreso = sesion.get(Egreso, egreso_id)
    if egreso is None:
        raise HTTPException(404, "Egreso no encontrado.")
    return egreso


@router.post("/{egreso_id}/anular", response_model=EgresoSalida)
def anular_egreso(egreso_id: int, sesion: Session = Depends(obtener_sesion)):
    """
    Se anula, no se borra: la caja tiene que poder explicar cada movimiento,
    incluidos los que se deshicieron.
    """
    egreso = sesion.get(Egreso, egreso_id)
    if egreso is None:
        raise HTTPException(404, "Egreso no encontrado.")
    if egreso.estado == "Anulado":
        raise HTTPException(409, "Este egreso ya está anulado.")

    egreso.estado = "Anulado"
    # Se vuelca antes de recalcular: la consulta de abajo excluye los anulados,
    # y sin esto seguiría contando este pago como vigente.
    sesion.flush()

    # El gasto vuelve a quedar pendiente si el pago que lo saldaba se anuló.
    if egreso.gasto_id:
        gasto = sesion.get(Gasto, egreso.gasto_id)
        if gasto is not None:
            pagado = sesion.scalar(
                select(func.coalesce(func.sum(Egreso.monto), 0)).where(
                    Egreso.gasto_id == gasto.id, Egreso.estado != "Anulado"
                )
            )
            pagado = redondear(pagado or 0)
            if pagado <= 0:
                gasto.estado_pago = "Por Pagar"
            elif pagado < gasto.total:
                gasto.estado_pago = "Parcial"

    sesion.commit()
    sesion.refresh(egreso)
    return egreso


# --------------------------------------------------------------------------
# Resumen
# --------------------------------------------------------------------------


@router.get("/resumen/periodo", response_model=ResumenEgresos)
def resumen(
    sesion: Session = Depends(obtener_sesion),
    desde: date | None = None,
    hasta: date | None = None,
):
    """Cuánto se gastó y cuánto se pagó en el período, y qué queda debiendo."""
    gastos = select(func.count(Gasto.id), func.coalesce(func.sum(Gasto.total), 0))
    pagos = select(func.coalesce(func.sum(Egreso.monto), 0)).where(Egreso.estado != "Anulado")

    if desde:
        gastos = gastos.where(Gasto.fecha >= desde)
        pagos = pagos.where(Egreso.fecha >= desde)
    if hasta:
        gastos = gastos.where(Gasto.fecha <= hasta)
        pagos = pagos.where(Egreso.fecha <= hasta)

    cantidad, total_gastos = sesion.execute(gastos).one()
    total_pagos = sesion.scalar(pagos)

    # Lo pendiente no se acota por período: una factura de hace meses sigue
    # debiéndose hoy, y filtrarla escondería justo la deuda más vieja.
    pendiente = sesion.scalar(
        select(func.coalesce(func.sum(Gasto.total), 0)).where(Gasto.estado_pago != "Pagado")
    )

    return ResumenEgresos(
        gastos=cantidad or 0,
        total_gastos=redondear(total_gastos or 0),
        total_pagos=redondear(total_pagos or 0),
        pendiente=redondear(pendiente or 0),
    )

File: frontend/src/api/egresos.js
/**
 * Egresos, anticipos y facturación recurrente contra el API.
 *
 * La distinción entre **gasto** y **egreso** es la del backend y conviene
 * repetirla aquí: el gasto es la obligación (llegó la factura del arriendo) y
 * el egreso es la salida de dinero (se pagó). No coinciden ni en fecha ni en
 * importe, por eso son dos listados distintos.
 */

import { api } from './cliente';

const numero = (valor) => Number(valor ?? 0);

export const FORMAS_PAGO_EGRESO = [
  'Efectivo',
  'Transferencia',
  'Cheque',
  'Tarjeta de crédito',
  'Tarjeta de débito',
];

export const ESTADOS_PAGO_GASTO = ['Por Pagar', 'Parcial', 'Pagado'];

export const PERIODICIDADES = [
  'Semanal',
  'Quincenal',
  'Mensual',
  'Bimestral',
  'Trimestral',
  'Anual',
];

/** ARD: recibido de un cliente. APP: pagado a un proveedor. */
export const TIPOS_ANTICIPO = [
  { codigo: 'ARD', nombre: 'Recibido de cliente' },
  { codigo: 'APP', nombre: 'Pagado a proveedor' },
];

export const ESTADOS_ANTICIPO = ['Pendiente', 'Parcial', 'Aplicado', 'Anulado'];

// --------------------------------------------------------------------------
// Tipos de gasto
// --------------------------------------------------------------------------

export const tipoGastoDesdeApi = (registro) => ({
  id: registro.id,
  nombre: registro.nombre,
  descripcion: registro.descripcion ?? '',
  deducible: registro.deducible,
  estado: registro.estado,
});

export const tipoGastoHaciaApi = (tipo) => ({
  nombre: tipo.nombre,
  descripcion: tipo.descripcion || null,
  // `?? true` y no `Boolean(...)`: con `Boolean` un campo sin tocar viajaría
  // como false e invertiría el valor por defecto del backend, marcando el
  // gasto como no deducible sin que nadie lo pidiera.
  deducible: tipo.deducible ?? true,
  estado: tipo.estado ?? 'Activo',
});

export const cargarTiposGasto = (opciones) =>
  api.obtener('/egresos/tipos', undefined, opciones);
export const crearTipoGasto = (tipo) => api.crear('/egresos/tipos', tipoGastoHaciaApi(tipo));
export const actualizarTipoGasto = (id, tipo) =>
  api.actualizar(`/egresos/tipos/${id}`, tipoGastoHaciaApi(tipo));
export const desactivarTipoGasto = (id) => api.eliminar(`/egresos/tipos/${id}`);

// --------------------------------------------------------------------------
// Gastos
// --------------------------------------------------------------------------

export const gastoDesdeApi = (registro) => ({
  id: registro.id,
  fecha: registro.fecha,
  concepto: registro.concepto,
  tipoId: registro.tipo_id,
  proveedor: registro.proveedor_razon_social,
  identificacion: registro.proveedor_identificacion,
  documento: registro.documento,
  fechaDocumento: registro.fecha_documento,
  autorizacionProveedor: registro.autorizacion_proveedor ?? '',
  subtotal: numero(registro.subtotal),
  iva: numero(registro.iva),
  codigoIva: registro.codigo_iva ?? '4',
  total: numero(registro.total),
  estadoPago: registro.estado_pago,
  observacion: registro.observacion ?? '',
});

export const gastoHaciaApi = (gasto) => ({
  fecha: gasto.fecha || null,
  concepto: gasto.concepto,
  tipo_id: gasto.tipoId ? Number(gasto.tipoId) : null,
  proveedor_id: gasto.proveedorId ? Number(gasto.proveedorId) : null,
  documento: gasto.documento || '',
  fecha_documento: gasto.fechaDocumento || null,
  autorizacion_proveedor: gasto.autorizacionProveedor || null,
  subtotal: String(gasto.subtotal ?? '0'),
  iva: String(gasto.iva ?? '0'),
  codigo_iva: gasto.codigoIva || '4',
  estado_pago: gasto.estadoPago ?? 'Por Pagar',
  observacion: gasto.observacion || null,
});

export const crearGasto = (gasto) => api.crear('/egresos/gastos', gastoHaciaApi(gasto));
export const actualizarGasto = (id, gasto) =>
  api.actualizar(`/egresos/gastos/${id}`, gastoHaciaApi(gasto));
export const eliminarGasto = (id) => api.eliminar(`/egresos/gastos/${id}`);

// --------------------------------------------------------------------------
// Egresos (pagos)
// --------------------------------------------------------------------------

export const egresoDesdeApi = (registro) => ({
  id: registro.id,
  fecha: registro.fecha,
  concepto: registro.concepto,
  beneficiario: registro.beneficiario,
  monto: numero(registro.monto),
  formaPago: registro.forma_pago,
  referencia: registro.referencia ?? '',
  gastoId: registro.gasto_id,
  estado: registro.estado,
  observacion: registro.observacion ?? '',
});

export const egresoHaciaApi = (egreso) => ({
  fecha: egreso.fecha || null,
  concepto: egreso.concepto,
  beneficiario: egreso.beneficiario || '',
  monto: String(egreso.monto ?? '0'),
  forma_pago: egreso.formaPago ?? 'Efectivo',
  cuenta_id: egreso.cuentaId ? Number(egreso.cuentaId) : null,
  referencia: egreso.referencia || null,
  gasto_id: egreso.gastoId ? Number(egreso.gastoId) : null,
  observacion: egreso.observacion || null,
});

export const crearEgreso = (egreso) => api.crear('/egresos', egresoHaciaApi(egreso));
export const anularEgreso = (id) => api.crear(`/egresos/${id}/anular`);

export const cargarResumenEgresos = (desde, hasta, opciones) =>
  api.obtener('/egresos/resumen/periodo', { desde, hasta }, opciones);

// --------------------------------------------------------------------------
// Anticipos
// --------------------------------------------------------------------------

export const anticipoDesdeApi = (registro) => ({
  id: registro.id,
  fecha: registro.fecha,
  tipo: registro.tipo,
  receptor: registro.receptor_razon_social,
  detalle: registro.detalle,
  monto: numero(registro.monto),
  facturado: numero(registro.facturado),
  saldo: numero(registro.saldo),
  formaPago: registro.forma_pago,
  estado: registro.estado,
});

export const anticipoHaciaApi = (anticipo, receptorId) => ({
  fecha: anticipo.fecha || null,
  tipo: anticipo.tipo ?? 'ARD',
  receptor_id: receptorId,
  detalle: anticipo.detalle || '',
  monto: String(anticipo.monto ?? '0'),
  forma_pago: anticipo.formaPago ?? 'Transferencia',
});

export const crearAnticipo = (anticipo, receptorId) =>
  api.crear('/anticipos', anticipoHaciaApi(anticipo, receptorId));
export const aplicarAnticipo = (id, monto) =>
  api.crear(`/anticipos/${id}/aplicar`, { monto: String(monto) });
export const anularAnticipo = (id) => api.crear(`/anticipos/${id}/anular`);
export const eliminarAnticipo = (id) => api.eliminar(`/anticipos/${id}`);

// --------------------------------------------------------------------------
// Facturación recurrente
// --------------------------------------------------------------------------

export const plantillaDesdeApi = (registro) => ({
  id: registro.id,
  nombre: registro.nombre,
  receptor: registro.receptor_razon_social,
  periodicidad: registro.periodicidad,
  proximaEmision: registro.proxima_emision,
  ultimaEmision: registro.ultima_emision,
  hasta: registro.hasta,
  total: numero(registro.total),
  emitidas: registro.emitidas,
  activa: registro.activa,
  lineas: (registro.lineas ?? []).map((linea) => ({
    id: linea.id,
    codigo: linea.codigo_principal,
    descripcion: linea.descripcion,
    cantidad: numero(linea.cantidad),
    precioUnitario: numero(linea.precio_unitario),
    codigoIva: linea.codigo_iva,
  })),
});

export const plantillaHaciaApi = (plantilla, receptorId, lineas) => ({
  nombre: plantilla.nombre,
  receptor_id: receptorId,
  periodicidad: plantilla.periodicidad ?? 'Mensual',
  proxima_emision: plantilla.proximaEmision,
  hasta: plantilla.hasta || null,
  establecimiento: '001',
  punto_emision: '001',
  forma_pago: plantilla.formaPago ?? '01',
  activa: plantilla.activa ?? true,
  lineas: lineas.map((linea) => ({
    codigo_principal: linea.codigo || 'SIN-COD',
    descripcion: linea.descripcion,
    cantidad: String(linea.cantidad ?? '1'),
    precio_unitario: String(linea.precioUnitario ?? '0'),
    descuento_porcentaje: String(linea.descuentoPorcentaje ?? '0'),
    codigo_iva: linea.codigoIva ?? '4',
  })),
});

export const crearPlantilla = (plantilla, receptorId, lineas) =>
  api.crear('/recurrentes', plantillaHaciaApi(plantilla, receptorId, lineas));
export const actualizarPlantilla = (id, plantilla, receptorId, lineas) =>
  api.actualizar(`/recurrentes/${id}`, plantillaHaciaApi(plantilla, receptorId, lineas));
export const pausarPlantilla = (id) => api.crear(`/recurrentes/${id}/pausar`);
export const emitirPlantilla = (id) => api.crear(`/recurrentes/${id}/emitir`);
export const eliminarPlantilla = (id) => api.eliminar(`/recurrentes/${id}`);
export const cargarVencidas = (opciones) =>
  api.obtener('/recurrentes/vencidas', undefined, opciones);
